src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs

```diff
@@ -233,11 +233,17 @@ private void ReplayWarningsAndErrors()
         private KeyValuePair<CacheFile,bool> EvaluateCacheFile()
         {
             CacheFile cacheFile;
-            var newDgSpecHash = _request.DependencyGraphSpec.GetHash();
             var noOp = false;
+
+            var newDgSpecHash = NoOpRestoreUtilities.GetHash(_request);
+
+            if(_request.ProjectStyle == ProjectStyle.DotnetCliTool && _request.AllowNoOp) { // No need to attempt to resolve the tool if no-op is not allowed.
+                NoOpRestoreUtilities.ResolveBestMatchingToolPathIfAvailable(_request);
+            }
+
             if (_request.AllowNoOp && File.Exists(_request.Project.RestoreMetadata.CacheFilePath))
             {
-                cacheFile = CacheFileFormat.Load(_request.Project.RestoreMetadata.CacheFilePath, _logger);
+                cacheFile = FileUtility.SafeRead(_request.Project.RestoreMetadata.CacheFilePath, _logger, (stream, log, path) => CacheFileFormat.Read(stream, log, path));
```

Comment:
It might make more sense to remove ILogger from SafeRead and pass it through the Func using the current local context. 

``cacheFile = FileUtility.SafeRead(_request.Project.RestoreMetadata.CacheFilePath, (stream, path) => CacheFileFormat.Read(stream, _logger, path));``